# Shibir Chat — RAG Corpus Analysis (Bengali / Arabic / English)

Fresh analysis of the production corpus (Postgres `pages` + `articles`) for RAG
architecture decisions: script/language mix, chunking behavior under the real
`app.rag.chunker.chunk_text()`, and data-quality issues that affect retrieval
(duplicate content, HTML residue, degenerate rows).

Everything below reads the **live production Postgres DB** and, in the
duplicate-content section, the **live Chroma index** (`documents_bge_m3`) —
no cached exports, no SQLite source dumps. Run top to bottom; each section is
self-contained given the setup cell.

In [1]:
import re
import statistics

import pandas as pd
from sqlalchemy import text

from app.core.config import settings
from app.db.session import engine
from app.rag.chunker import chunk_text

pd.set_option("display.width", 140)
pd.set_option("display.max_colwidth", 120)

with engine.connect() as conn:
    pages_df = pd.read_sql_query(
        text(
            "SELECT id, book_id, chapter_id, language, status, content, "
            "excluded_from_rag, exclusion_reason, source_db, embedded_at "
            "FROM pages"
        ),
        conn,
    )
    articles_df = pd.read_sql_query(
        text(
            "SELECT id, title, language, status, content, source_type, embedded_at "
            "FROM articles"
        ),
        conn,
    )
    books_df = pd.read_sql_query(text("SELECT id, name, category_id FROM books"), conn)

print(f"connected: {engine.url.render_as_string(hide_password=True)}")
print(f"pages: {len(pages_df)}   articles: {len(articles_df)}   books: {len(books_df)}")

connected: postgresql+psycopg2://postgres:***@localhost:5432/shibir_chat
pages: 4143   articles: 234   books: 55


## 1. Corpus overview

What's actually eligible for the RAG index right now.

In [2]:
print("pages.status:", pages_df["status"].value_counts().to_dict())
print("articles.status:", articles_df["status"].value_counts().to_dict())
print(f"pages.excluded_from_rag = True: {pages_df['excluded_from_rag'].sum()} "
      f"(soft-excluded by scripts/clean_corpus.py)")
print(f"pages embedded_at set: {pages_df['embedded_at'].notna().sum()}/{len(pages_df)}")
print(f"articles embedded_at set: {articles_df['embedded_at'].notna().sum()}/{len(articles_df)}")

# The active RAG corpus: published + not soft-excluded.
active_pages = pages_df[(pages_df["status"] == "published") & (~pages_df["excluded_from_rag"])].copy()
active_articles = articles_df[articles_df["status"] == "published"].copy()
print(f"\nactive pages: {len(active_pages)}   active articles: {len(active_articles)}   "
      f"total active docs: {len(active_pages) + len(active_articles)}")

pages.status: {'published': 4143}
articles.status: {'published': 234}
pages.excluded_from_rag = True: 106 (soft-excluded by scripts/clean_corpus.py)
pages embedded_at set: 4143/4143
articles embedded_at set: 234/234

active pages: 4037   active articles: 234   total active docs: 4271


## 2. Script / language mix (char-level)

This corpus is not monolingual Bengali — Qur'anic ayat and hadith are quoted
in Arabic script inline, and some pages mix in Latin (English loanwords,
transliteration). This directly informs the embedding model choice
(`BAAI/bge-m3`, multilingual, replacing `all-MiniLM-L6-v2` which can't
tokenize Bengali at all) and the reranker's cross-lingual burden.

In [3]:
BENGALI_RE = re.compile(r"[\u0980-\u09FF]")
ARABIC_RE = re.compile(r"[\u0600-\u06FF]")
LATIN_RE = re.compile(r"[A-Za-z]")


def script_profile(t: str) -> dict:
    n = len(t) or 1
    return {
        "bn_frac": len(BENGALI_RE.findall(t)) / n,
        "ar_frac": len(ARABIC_RE.findall(t)) / n,
        "en_frac": len(LATIN_RE.findall(t)) / n,
    }


def dominant_script(p: dict) -> str:
    bn, ar, en = p["bn_frac"], p["ar_frac"], p["en_frac"]
    if max(bn, ar, en) < 0.05:
        return "other/numeric/symbols"
    ordered = sorted([bn, ar, en], reverse=True)
    if ordered[0] > 0 and ordered[1] / ordered[0] > 0.35:
        return "mixed"
    return {bn: "bn", ar: "ar", en: "en"}[ordered[0]]


for label, df in [("pages", active_pages), ("articles", active_articles)]:
    profiles = df["content"].fillna("").apply(script_profile)
    dom = profiles.apply(dominant_script)
    dist = dom.value_counts(normalize=True).mul(100).round(1)
    mean_ar = profiles.apply(lambda p: p["ar_frac"]).mean() * 100
    print(f"{label} (n={len(df)}) — dominant script (%):")
    print(dist.to_string())
    print(f"  mean Arabic-char share across all {label}: {mean_ar:.2f}%\n")

pages (n=4037) — dominant script (%):
content
bn       93.6
mixed     6.4
  mean Arabic-char share across all pages: 3.60%

articles (n=234) — dominant script (%):
content
bn    100.0
  mean Arabic-char share across all articles: 0.00%



## 3. Raw content length (pre-chunk)

In [4]:
for label, df in [("pages", active_pages), ("articles", active_articles)]:
    lens = df["content"].fillna("").str.len()
    print(
        f"{label}: n={len(df)} min={lens.min()} median={lens.median():.0f} "
        f"mean={lens.mean():.0f} p95={lens.quantile(0.95):.0f} max={lens.max()}"
    )

pages: n=4037 min=24 median=1738 mean=1876 p95=3670 max=18299
articles: n=234 min=324 median=631 mean=622 p95=757 max=860


## 4. Chunking simulation — real `app.rag.chunker.chunk_text`

Runs the exact function used at ingest time (900-char target, 150-char
overlap, paragraph → Bengali danda (`।`) → Latin period → space split
priority) against every active doc, so the numbers below match what's
actually sitting in the Chroma index today.

In [5]:
chunk_rows = []
tiny_chunks = []
all_chunk_rows = []

for label, df in [("pages", active_pages), ("articles", active_articles)]:
    lens, n_tiny = [], 0
    for rid, content in zip(df["id"], df["content"]):
        for i, c in enumerate(chunk_text(content or "")):
            lens.append(len(c))
            if len(c) < 50:
                n_tiny += 1
                tiny_chunks.append((label, rid, i, c))
            prof = script_profile(c)
            all_chunk_rows.append({"source": label, "doc_id": rid, "chunk_index": i, "len": len(c), **prof})
    chunk_rows.append({
        "source": label,
        "n_docs": len(df),
        "n_chunks": len(lens),
        "chunks_per_doc": round(len(lens) / max(len(df), 1), 2),
        "median_chunk_len": statistics.median(lens) if lens else 0,
        "p95_chunk_len": sorted(lens)[int(len(lens) * 0.95)] if lens else 0,
        "n_tiny_lt50": n_tiny,
    })

chunks_summary = pd.DataFrame(chunk_rows)
all_chunks_df = pd.DataFrame(all_chunk_rows)
print(chunks_summary.to_string(index=False))
print(f"\ntotal chunks across corpus: {chunks_summary['n_chunks'].sum()}")

  source  n_docs  n_chunks  chunks_per_doc  median_chunk_len  p95_chunk_len  n_tiny_lt50
   pages    4037     12070            2.99             779.0           1032           42
articles     234       234            1.00             640.5            770            0

total chunks across corpus: 12304


In [6]:
print(f"tiny chunks (<50 chars): {len(tiny_chunks)} — isolated headings that couldn't merge forward\n")
for label, rid, i, c in tiny_chunks[:10]:
    print(f"  [{label}] id={rid} chunk={i}: {c!r}")

tiny chunks (<50 chars): 42 — isolated headings that couldn't merge forward

  [pages] id=5459 chunk=0: '‘ইলাহ’ শব্দের অর্থ'
  [pages] id=5485 chunk=0: '‘ইলাহ’ শব্দের অর্থ'
  [pages] id=5603 chunk=0: 'খোদার সাথে সম্পর্ক ও আন্তরিকতা'
  [pages] id=5605 chunk=0: 'চরিত্র মাধুর্য'
  [pages] id=5611 chunk=0: 'মৌলিক ও অসৎ গুনাবলী\n\nগর্ব ও অহংকার'
  [pages] id=5821 chunk=0: 'মু’জিযা ও মি’রাজ'
  [pages] id=5840 chunk=0: 'মসজিদে নববীর নির্মাণ'
  [pages] id=5904 chunk=0: 'হুনাইনের যুদ্ধ\n\nমক্কা বিজয়ের প্রভাব'
  [pages] id=5922 chunk=0: 'বিদায় হজ্জ এবং ওফাত\n\nহজ্জের জন্য রওয়ানা'
  [pages] id=5986 chunk=0: 'সা’দ ইবন আবী ওয়াক্কাস (রা)'


### 4.1 All chunks (full table)

Every chunk that gets embedded, across both sources — `chunk_text` preview
truncated to 300 chars for display only; nothing in the underlying dataframe
is truncated (`all_chunks_full_df.iloc[i]['chunk_text']` has the full text).

In [7]:
all_chunks_full_rows = []
for label, df in [("pages", active_pages), ("articles", active_articles)]:
    for rid, content in zip(df["id"], df["content"]):
        for i, c in enumerate(chunk_text(content or "")):
            all_chunks_full_rows.append({
                "source": label, "doc_id": rid, "chunk_index": i,
                "chunk_len": len(c), "chunk_text": c,
            })
all_chunks_full_df = pd.DataFrame(all_chunks_full_rows)
print(f"total chunks: {len(all_chunks_full_df)} "
      f"(pages: {(all_chunks_full_df['source']=='pages').sum()}, "
      f"articles: {(all_chunks_full_df['source']=='articles').sum()})")

preview = all_chunks_full_df.copy()
preview["chunk_text"] = preview["chunk_text"].str.slice(0, 300)
preview.head(30)

total chunks: 12304 (pages: 12070, articles: 234)


,source,doc_id,chunk_index,chunk_len,chunk_text
0,pages,4260,0,523,ইবারত\n\nوَمَا مُحَمَّدٌ إِلَّا رَسُولٌ قَدْ خَلَتْ مِن قَبْلِهِ الرُّسُلُ ۚ أَفَإِن مَّاتَ أَوْ قُتِلَ انقَلَبْتُمْ...
1,pages,4260,1,938,"াবে ? মনে রেখো, যে পেছনের দিকে ফিরে যাবে সে আল্লাহর কোন ক্ষতি করবে না, তবে যারা আল্লাহর কৃতজ্ঞ বান্দা হয়ে থাকবে তাদ..."
2,pages,4260,2,652,"য় নেবার সাথে সাথে তোমরা আবার সেই কুফরীর দিকে ফিরে যাবে, যা থেকে তোমরা বের হয়ে এসেছিলে, তাহলে আল্লাহর দীন তোমাদের ক..."
3,pages,4260,3,553,য়া থেকেই দেবো৷ আর যে ব্যক্তি পরকালীন পুরস্কার১০৫ লাভের আশায় কাজ করবে সে পরকালের পুরস্কার পাবে এবং শোকরকারীদেরকে১০৬...
4,pages,4260,4,659,্তা না করে বরং জীবিত থাকার জন্য যে সময়টুকু পাচ্ছো সেই সময়ে তোমাদের যাবতীয় প্রচেষ্টার উদ্দেশ্যে দুনিয়া না আখেরাত ...
5,pages,4260,5,1047,"দৃষ্টি ইহকালীন না পরকালীন ফল প্রাপ্তির দিকে নিবন্ধ থাকবে, ইসলামের দৃষ্টিতে মানবিক নৈতিকতার ক্ষেত্রে এটিই হচ্ছে চূড়া..."
6,pages,4260,6,528,"আল্লাহ তাকে এ ব্যাপারে নিশ্চয়তা দিয়েছেন যে, আখেরাতে সে অবিশ্য এর ভালো ফল পাবে- এহেন ব্যক্তিই হচ্ছে আল্লাহর শোকর গু..."
7,pages,4260,7,999,"য় করতে প্রস্তুত হয় না, তারাই সত্যিকার অর্থে না-শোকরগুজার ও অকৃতজ্ঞ বান্দা৷ আল্লাহ তাদেরকে যে জ্ঞান দান করেছেন তাদে..."
8,pages,5246,0,303,"শিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ইসলামের ধারণা বা দৃষ্টিকোণ কি, তার সংজ্ঞ..."
9,pages,5246,1,1030,ড়ান্ত ও স্থির সিদ্ধান্তে পৌঁছতে হলে প্রথমেই মানুষ সম্পর্কে ইসলামের ধারণা ও দৃষ্টিকোণকে মৌলিকভাবে বিচার-বিবেচনা করতে...


## 5. Arabic-bearing chunks — how much cross-lingual work the reranker does

Ayat/hadith quoted inside Bengali tafsir prose produce chunks that are
partly or mostly Arabic script. A Bengali query matching such a chunk relies
on `bge-m3` / `bge-reranker-v2-m3`'s cross-lingual alignment, not lexical
overlap — worth its own eval slice, since these aren't rare.

In [8]:
all_chunks_df["dominant"] = all_chunks_df.apply(
    lambda r: dominant_script({"bn_frac": r.bn_frac, "ar_frac": r.ar_frac, "en_frac": r.en_frac}),
    axis=1,
)
print("dominant script per chunk (%):")
print(all_chunks_df["dominant"].value_counts(normalize=True).mul(100).round(1).to_string())

n_arabic_heavy = (all_chunks_df["ar_frac"] > 0.5).sum()
n_arabic_any = (all_chunks_df["ar_frac"] > 0.05).sum()
print(f"\nchunks >50% Arabic characters: {n_arabic_heavy} ({n_arabic_heavy/len(all_chunks_df)*100:.1f}%)")
print(f"chunks with >5% Arabic characters (verse embedded in bn prose): "
      f"{n_arabic_any} ({n_arabic_any/len(all_chunks_df)*100:.1f}%)")

dominant script per chunk (%):
dominant
bn       92.2
mixed     7.3
ar        0.4
en        0.0

chunks >50% Arabic characters: 120 (1.0%)
chunks with >5% Arabic characters (verse embedded in bn prose): 1704 (13.8%)


## 6. HTML residue check

The chunker does not strip HTML — any tag that survived `clean_html` at migration time goes straight into the embedding.

In [9]:
HTML_RE = re.compile(r"<[a-zA-Z][^>]*>")
for label, df in [("pages", active_pages), ("articles", active_articles)]:
    has_html = df["content"].fillna("").str.contains(HTML_RE)
    print(f"{label}: {has_html.sum()}/{len(df)} rows ({has_html.mean()*100:.2f}%) contain raw HTML tags")
    if has_html.sum():
        sample = df.loc[has_html, ["id", "content"]].head(3)
        for _, row in sample.iterrows():
            print(f"    id={row['id']}: {row['content'][:150]!r}")

pages: 3/4037 rows (0.07%) contain raw HTML tags
    id=8088: '<p><strong>&nbsp;</strong></p>\n<p><strong>একটি হাদীস এবং আববকর</strong><strong> (রা)</strong></p>\n<p>এই হাদীস শুনানোর পরবর্তীকালে হযরত উমার (রা) বলেছি'
    id=7854: '<p style="text-align: justify;"><strong>প্রহারকারী স্বামীরা সর্বোত্তম মানুষ নয়</strong></p>\n<p style="text-align: justify;">(আরবী*******************)<'
    id=8327: '<p style="text-align: justify;">সহযোগিতায় রচনা করেছেন বর্তমান যুগের বিজ্ঞান জগতের সবচেয়ে আলোড়ন সষ্টিকারী গ্রন্থ &lsquo;এ ব্রীফ হিস্ট্রি অব টাইম &rsquo'
articles: 0/234 rows (0.00%) contain raw HTML tags


## 7. Duplicate-book check — cross-referenced against the *live* Chroma index

`scripts/clean_corpus.py` already removed a different problem (book203
cross-contamination). This section checks a separate issue: books that
share the same name (same source book imported twice under different
`book_id`s) and whether **both copies are still being served to users
today**, not just present in Postgres.

In [10]:
book_name_counts = books_df["name"].value_counts()
dup_names = book_name_counts[book_name_counts > 1].index.tolist()
print(f"book names appearing more than once: {len(dup_names)}")

dup_book_ids = books_df[books_df["name"].isin(dup_names)][["id", "name"]].sort_values("name")

with engine.connect() as conn:
    page_counts = pd.read_sql_query(
        text(
            "SELECT book_id, "
            "count(*) FILTER (WHERE NOT excluded_from_rag) AS active_pages, "
            "count(*) FILTER (WHERE excluded_from_rag) AS excluded_pages "
            f"FROM pages WHERE book_id IN ({','.join(map(str, dup_book_ids['id']))}) "
            "GROUP BY book_id"
        ),
        conn,
    )

dup_overview = dup_book_ids.merge(page_counts, left_on="id", right_on="book_id").drop(columns="book_id")
dup_overview["exact_pair"] = dup_overview.groupby("name")["active_pages"].transform(lambda s: s.nunique() == 1)
dup_overview.sort_values(["name", "id"])

book names appearing more than once: 6


,id,name,active_pages,excluded_pages,exact_pair
0,177,ঈমানের হাকীকত,41,0,True
1,211,ঈমানের হাকীকত,41,0,True
2,204,উলুমুল কুরআন ও উলুমুল হাদীস,24,0,True
3,219,উলুমুল কুরআন ও উলুমুল হাদীস,24,0,True
4,173,কর্মপদ্ধতি,45,0,False
5,210,কর্মপদ্ধতি,47,0,False
6,166,তাফহীমুল কুরআন,233,0,False
7,206,তাফহীমুল কুরআন,6,0,False
8,180,নামাজ-রোজার হাকীকত,59,0,True
9,212,নামাজ-রোজার হাকীকত,59,0,True


In [11]:
import chromadb

client = chromadb.PersistentClient(path=settings.chroma_persist_dir)
coll = client.get_collection(settings.chroma_collection_name)
print(f"live Chroma collection '{settings.chroma_collection_name}': {coll.count()} vectors")

metas = coll.get(include=["metadatas"])["metadatas"]
row_keys_in_chroma = {m.get("row_key") for m in metas if m}

with engine.connect() as conn:
    dup_pages = pd.read_sql_query(
        text(
            "SELECT id, book_id FROM pages "
            f"WHERE book_id IN ({','.join(map(str, dup_book_ids['id']))})"
        ),
        conn,
    )
dup_pages["row_key"] = "page_" + dup_pages["id"].astype(str)
dup_pages["in_chroma"] = dup_pages["row_key"].isin(row_keys_in_chroma)

print(f"\nduplicate-book pages currently embedded in Chroma: "
      f"{dup_pages['in_chroma'].sum()}/{len(dup_pages)}")

n_dup_chunks = sum(
    len(chunk_text(c or ""))
    for c in pages_df.loc[pages_df["book_id"].isin(dup_book_ids["id"]), "content"]
)
print(f"chunks belonging to any duplicate-named book: {n_dup_chunks} "
      f"({n_dup_chunks / chunks_summary['n_chunks'].sum() * 100:.1f}% of the whole index)")

exact_pair_ids = dup_overview.loc[dup_overview["exact_pair"], "id"].tolist()
wasted = 0
for name, grp in dup_overview[dup_overview["exact_pair"]].groupby("name"):
    ids = grp["id"].tolist()
    one_copy_content = pages_df.loc[pages_df["book_id"] == ids[0], "content"]
    wasted += sum(len(chunk_text(c or "")) for c in one_copy_content)
print(f"fully redundant chunks (one full copy of each EXACT-duplicate pair): {wasted}")

live Chroma collection 'documents_bge_m3': 12304 vectors



duplicate-book pages currently embedded in Chroma: 595/595


chunks belonging to any duplicate-named book: 2031 (16.5% of the whole index)
fully redundant chunks (one full copy of each EXACT-duplicate pair): 347


## 8. Empty / degenerate content

In [12]:
for label, df in [("pages", active_pages), ("articles", active_articles)]:
    stripped = df["content"].fillna("").str.strip()
    empty = (stripped.str.len() == 0).sum()
    very_short = ((stripped.str.len() > 0) & (stripped.str.len() < 20)).sum()
    print(f"{label}: empty={empty} very_short(<20 chars)={very_short}")

pages: empty=0 very_short(<20 chars)=0


articles: empty=0 very_short(<20 chars)=0


## 9. Export full dataset (CSV)

Every chunk that gets embedded (pages + articles), enriched with book/chapter/
article metadata and the per-chunk script profile from Section 5, written out
as one flat CSV — `full_dataset.csv` at the repo root. This is the same
`all_chunks_full_df` / `all_chunks_df` computed above, joined and saved, not a
re-derivation, so it stays consistent with every number reported earlier in
this notebook.

In [13]:
with engine.connect() as conn:
    chapters_df = pd.read_sql_query(text("SELECT id, name FROM chapters"), conn)

book_names = books_df.set_index("id")["name"]
chapter_names = chapters_df.set_index("id")["name"]

pages_meta = active_pages[["id", "book_id", "chapter_id", "language"]].rename(columns={"id": "doc_id"})
pages_meta["source"] = "pages"
pages_meta["book_name"] = pages_meta["book_id"].map(book_names)
pages_meta["chapter_name"] = pages_meta["chapter_id"].map(chapter_names)
pages_meta["article_title"] = None

articles_meta = active_articles[["id", "language", "title"]].rename(columns={"id": "doc_id", "title": "article_title"})
articles_meta["source"] = "articles"
articles_meta["book_id"] = None
articles_meta["book_name"] = None
articles_meta["chapter_id"] = None
articles_meta["chapter_name"] = None

doc_meta = pd.concat([pages_meta, articles_meta], ignore_index=True)

full_dataset_df = (
    all_chunks_full_df
    .merge(all_chunks_df[["source", "doc_id", "chunk_index", "bn_frac", "ar_frac", "en_frac", "dominant"]],
           on=["source", "doc_id", "chunk_index"], how="left")
    .merge(doc_meta, on=["source", "doc_id"], how="left")
)

full_dataset_df = full_dataset_df[[
    "source", "doc_id", "book_id", "book_name", "chapter_id", "chapter_name",
    "article_title", "language", "chunk_index", "chunk_len",
    "bn_frac", "ar_frac", "en_frac", "dominant", "chunk_text",
]].rename(columns={"dominant": "dominant_script"})

out_path = "full_dataset.csv"
full_dataset_df.to_csv(out_path, index=False, encoding="utf-8-sig")

import os
print(f"wrote {len(full_dataset_df)} rows, {len(full_dataset_df.columns)} columns -> {out_path} "
      f"({os.path.getsize(out_path) / 1024 / 1024:.2f} MB)")
full_dataset_df.head(10)

wrote 12304 rows, 15 columns -> full_dataset.csv (24.44 MB)


,source,doc_id,book_id,book_name,chapter_id,chapter_name,article_title,language,chunk_index,chunk_len,bn_frac,ar_frac,en_frac,dominant_script,chunk_text
0,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,0,523,0.435946,0.382409,0.000000,mixed,ইবারত\n\nوَمَا مُحَمَّدٌ إِلَّا رَسُولٌ قَدْ خَلَتْ مِن قَبْلِهِ الرُّسُلُ ۚ أَفَإِن مَّاتَ أَوْ قُتِلَ انقَلَبْتُمْ...
1,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,1,938,0.808102,0.000000,0.000000,bn,"াবে ? মনে রেখো, যে পেছনের দিকে ফিরে যাবে সে আল্লাহর কোন ক্ষতি করবে না, তবে যারা আল্লাহর কৃতজ্ঞ বান্দা হয়ে থাকবে তাদ..."
2,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,2,652,0.559816,0.274540,0.000000,mixed,"য় নেবার সাথে সাথে তোমরা আবার সেই কুফরীর দিকে ফিরে যাবে, যা থেকে তোমরা বের হয়ে এসেছিলে, তাহলে আল্লাহর দীন তোমাদের ক..."
3,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,3,553,0.822785,0.000000,0.000000,bn,য়া থেকেই দেবো৷ আর যে ব্যক্তি পরকালীন পুরস্কার১০৫ লাভের আশায় কাজ করবে সে পরকালের পুরস্কার পাবে এবং শোকরকারীদেরকে১০৬...
4,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,4,659,0.827011,0.000000,0.000000,bn,্তা না করে বরং জীবিত থাকার জন্য যে সময়টুকু পাচ্ছো সেই সময়ে তোমাদের যাবতীয় প্রচেষ্টার উদ্দেশ্যে দুনিয়া না আখেরাত ...
5,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,5,1047,0.824260,0.000000,0.000000,bn,"দৃষ্টি ইহকালীন না পরকালীন ফল প্রাপ্তির দিকে নিবন্ধ থাকবে, ইসলামের দৃষ্টিতে মানবিক নৈতিকতার ক্ষেত্রে এটিই হচ্ছে চূড়া..."
6,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,6,528,0.827652,0.000000,0.000000,bn,"আল্লাহ তাকে এ ব্যাপারে নিশ্চয়তা দিয়েছেন যে, আখেরাতে সে অবিশ্য এর ভালো ফল পাবে- এহেন ব্যক্তিই হচ্ছে আল্লাহর শোকর গু..."
7,pages,4260,166,তাফহীমুল কুরআন,2167,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,bn,7,999,0.524525,0.308308,0.000000,mixed,"য় করতে প্রস্তুত হয় না, তারাই সত্যিকার অর্থে না-শোকরগুজার ও অকৃতজ্ঞ বান্দা৷ আল্লাহ তাদেরকে যে জ্ঞান দান করেছেন তাদে..."
8,pages,5246,190,শিক্ষা সাহিত্য সংস্কৃতি,2362,শিক্ষা,None,bn,0,303,0.838284,0.000000,0.000000,bn,"শিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ইসলামের ধারণা বা দৃষ্টিকোণ কি, তার সংজ্ঞ..."
9,pages,5246,190,শিক্ষা সাহিত্য সংস্কৃতি,2362,শিক্ষা,None,bn,1,1030,0.812621,0.000000,0.005825,bn,ড়ান্ত ও স্থির সিদ্ধান্তে পৌঁছতে হলে প্রথমেই মানুষ সম্পর্কে ইসলামের ধারণা ও দৃষ্টিকোণকে মৌলিকভাবে বিচার-বিবেচনা করতে...


## 10. Export full PostgreSQL dataset (CSV)

Unlike Section 9 (chunked, RAG-active-only), this is the **raw source-of-truth
dump** straight from Postgres: every `pages` row and every `articles` row,
unfiltered — published or not, excluded from RAG or not, unchunked content.
`pages` is joined with `books`/`categories`/`chapters` for readable names;
`articles` keeps its own columns (`title`, `author`, `source_type`, ...).
The two record types are stacked into one long table with a `record_type`
column and written to `postgres_full_dataset.csv` at the repo root.

In [14]:
with engine.connect() as conn:
    categories_df = pd.read_sql_query(text("SELECT id, name AS category_name FROM categories"), conn)

books_full = books_df.rename(columns={"name": "book_name"}).merge(
    categories_df, left_on="category_id", right_on="id", suffixes=("", "_cat")
)[["id", "book_name", "category_id", "category_name"]]

pages_raw = pages_df.merge(
    books_full, left_on="book_id", right_on="id", how="left", suffixes=("", "_book")
).merge(
    chapters_df.rename(columns={"name": "chapter_name"}), left_on="chapter_id", right_on="id",
    how="left", suffixes=("", "_chapter"),
)
pages_raw["record_type"] = "page"
pages_raw = pages_raw.rename(columns={"language": "language", "content": "content"})

articles_raw = articles_df.copy()
articles_raw["record_type"] = "article"

common_cols = [
    "record_type", "id", "book_name", "category_name", "chapter_name", "title", "author",
    "language", "status", "content", "excluded_from_rag", "exclusion_reason", "source_db",
    "source_page_id", "source_type", "source_ref", "source_metadata",
    "created_at", "updated_at", "embedded_at", "published_at",
]
for col in common_cols:
    if col not in pages_raw.columns:
        pages_raw[col] = None
    if col not in articles_raw.columns:
        articles_raw[col] = None

postgres_full_df = pd.concat([pages_raw[common_cols], articles_raw[common_cols]], ignore_index=True)
postgres_full_df["source_metadata"] = postgres_full_df["source_metadata"].astype(str).where(
    postgres_full_df["source_metadata"].notna(), None
)

out_path = "postgres_full_dataset.csv"
postgres_full_df.to_csv(out_path, index=False, encoding="utf-8-sig")

import os
print(f"wrote {len(postgres_full_df)} rows, {len(postgres_full_df.columns)} columns -> {out_path} "
      f"({os.path.getsize(out_path) / 1024 / 1024:.2f} MB)")
print(f"  pages: {(postgres_full_df['record_type']=='page').sum()} "
      f"(excluded_from_rag={postgres_full_df['excluded_from_rag'].sum()})")
print(f"  articles: {(postgres_full_df['record_type']=='article').sum()}")
postgres_full_df.head(10)

wrote 4377 rows, 21 columns -> postgres_full_dataset.csv (20.49 MB)
  pages: 4143 (excluded_from_rag=106)
  articles: 234


,record_type,id,book_name,category_name,chapter_name,title,author,language,status,content,...,exclusion_reason,source_db,source_page_id,source_type,source_ref,source_metadata,created_at,updated_at,embedded_at,published_at
0,page,4260,তাফহীমুল কুরআন,আল কুরআন,সূরা আলে ইমরান (১৩ - ২০ রুকু),None,None,bn,published,ইবারত\n\n\nوَمَا مُحَمَّدٌ إِلَّا رَسُولٌ قَدْ خَلَتْ مِن قَبْلِهِ الرُّسُلُ ۚ أَفَإِن مَّاتَ أَوْ قُتِلَ انقَلَبْتُ...,...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
1,page,5246,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"শিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ইসলামের ধারণা বা দৃষ্টিকোণ কি, তার সংজ...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
2,page,5247,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"এই পরিপ্রেক্ষিতে জ্ঞানের সংজ্ঞায় বলা যায়\n\n\nস্রষ্টার (আল্লাহর) অস্তিত্ব, তাঁর গুণাবলীর তত্ত্ব ও বাস্তবতা, বিশ্বলোক...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
3,page,5248,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,ইসলামের শিক্ষা–দর্শন\n\n\n১৯৫১ সালের নভেম্বর মাসে ইংল্যান্ডের শিক্ষা বিভাগের কর্মকর্তা মিঃ ঊড্ (Mr. Wood) বিশ্ব-পর্য...,...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
4,page,5249,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"‘‘জেনে রাখ, মানবদেহে এমন একটা মাংসপিণ্ড বর্তমান, যা সুস্থ হলে সমস্ত দেহ সুস্থ থাকে। আর তা রোগাক্রান্ত হলে সমস্ত দেহ ...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
5,page,5250,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"‘‘যারা পরকাল বিশ্বাস করে না, তারা ফেরেশতাদেরকে দেবীদের নামে অভিহিত করে। অথচ এ ব্যাপারে তাদের কিছুই জানা নেই। তারা তো...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
6,page,5251,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,‘‘এই দুনিয়ার জীবনটা তো কয়েক দিনের ক্ষয়িষ্ণু সামগ্রী মাত্র। আর চিরকাল অবস্থানে জায়গা তো হল পরকাল।’’\n (\nসূরা\n \nমুম...,...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
7,page,5252,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"‘‘আকাশমণ্ডলে যা-কিছু আছে আর যা-কিছু রয়েছে পৃথিবীতে, সেই সব কিছুকেই আল্লাহ তা’আলা (হে মানুষ! কেবল) তোমাদেরই কল্যাণে ও...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
8,page,5264,শিক্ষা সাহিত্য সংস্কৃতি,ইসলামী জীবনব্যবস্থা,শিক্ষা,None,None,bn,published,"তবে কুরআনে মানুষের বিশেষ সম্মান ও মর্যাদার কথা বলা হয়েছে। কিন্তু মানুষ সে সম্মান ও মর্যাদার অধিকারী হবে কেবল তখন, যখ...",...,NaN,tarun,None,None,None,NaN,None,None,2026-08-17 18:43:36.989071+00:00,None
9,page,5632,ইসলামী আন্দোলন সাফল্যের শর্তাবলী,ইসলামী আন্দোলন,মানবিক দুর্বলতা,None,None,bn,published,“সাবধান! তোমরা একদেশদর্শীতা ও চরম পন্থা অবলম্বন করো না। কারণ তোমাদের পূর্ববর্তীরা চরমপন্থা অবলম্বন করেই ধ্বংস হয়েছে।...,...,NaN,tarun,None,None,None,NaN,None,None,2026-08-20 11:52:21.585382+00:00,None


## 11. Post-embedding dataset — live Chroma index

Sections 9 and 10 export *pre-embedding* data (the notebook's own
`chunk_text()` simulation, and the raw Postgres rows). This section pulls
the **actual post-embedding state**: what's really sitting in the live
`documents_bge_m3` Chroma collection right now — the exact text that was
embedded (including the `বই: / অধ্যায়:` header `build_document()` prepends,
which Section 4's simulation did not add), its metadata, and the bge-m3
vectors themselves.

**Disk note:** this box is at 98% disk usage (~3 GB free at analysis time).
12,304 rows x 1024-dim float32 vectors as CSV text would run 200MB+ for no
real benefit (nobody eyeballs a vector in a spreadsheet). So: metadata + the
source text (renamed `source_text` -- it's the string that WAS embedded, not
the embedding itself) go to `chroma_embedded_dataset.csv` (same shape as the
earlier CSVs), and the vectors go to a compact binary `chroma_embeddings.npy`
(~50 MB, float32) — **row order matches the CSV exactly**, so
`chroma_embeddings.npy[i]` is the embedding for `chroma_embedded_dataset.csv`
row `i`.

In [15]:
import numpy as np

client = chromadb.PersistentClient(path=settings.chroma_persist_dir)
coll = client.get_collection(settings.chroma_collection_name)
print(f"live collection '{settings.chroma_collection_name}': {coll.count()} vectors")

result = coll.get(include=["documents", "metadatas", "embeddings"])
ids = result["ids"]
docs = result["documents"]
metas = result["metadatas"]
embeddings = np.asarray(result["embeddings"], dtype=np.float32)

print(f"pulled {len(ids)} ids, embedding shape {embeddings.shape}")

chroma_rows = []
for _id, doc, meta in zip(ids, docs, metas):
    row_key = meta.get("row_key", "")
    source = "pages" if row_key.startswith("page_") else "articles" if row_key.startswith("article_") else "unknown"
    chroma_rows.append({
        "id": _id,
        "source": source,
        "doc_id": meta.get("page_id"),
        "row_key": row_key,
        "chunk_index": meta.get("chunk_index"),
        "book": meta.get("book"),
        "chapter": meta.get("chapter"),
        "source_db": meta.get("source_db"),
        "source_text_len": len(doc),
        "source_text": doc,
    })

chroma_df = pd.DataFrame(chroma_rows)
chroma_df["embedding_l2_norm"] = np.linalg.norm(embeddings, axis=1)
chroma_df = chroma_df[[
    "id", "source", "doc_id", "row_key", "chunk_index", "book", "chapter",
    "source_db", "source_text_len", "embedding_l2_norm", "source_text",
]]
chroma_df.head(5)

live collection 'documents_bge_m3': 12304 vectors


pulled 12304 ids, embedding shape (12304, 1024)


,id,source,doc_id,row_key,chunk_index,book,chapter,source_db,source_text_len,embedding_l2_norm,source_text
0,page_5246_c0,pages,5246,page_5246,0,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,tarun,348,1.0,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nশিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা স...
1,page_5246_c1,pages,5246,page_5246,1,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,tarun,1075,1.0,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nড়ান্ত ও স্থির সিদ্ধান্তে পৌঁছতে হলে প্রথমেই মানুষ সম্পর্কে ইসলামের ...
2,page_5246_c2,pages,5246,page_5246,2,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,tarun,976,1.0,"বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nসর্বোপরি মানুষের রয়েছে ন্যায়-অন্যায়, ভাল-মন্দ ও পাপ-পূণ্য বোধ। এই..."
3,page_5246_c3,pages,5246,page_5246,3,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,tarun,1007,1.0,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nন্ত্র্য আছে? সে নিজের বিবেচনায়ই যখন সর্বক্ষেত্রে নিজের সুস্পষ্ট স্ব...
4,page_5246_c4,pages,5246,page_5246,4,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,tarun,453,1.0,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nজবাব দেয়ারও কোন প্রয়োজন নেই। এই সব প্রশ্ন জাগে কেবল মানুষের মনে- ম...


In [16]:
print("source_text length (WITH header) stats:")
print(chroma_df["source_text_len"].describe())

merged_len_check = chroma_df.merge(
    all_chunks_full_df.rename(columns={"chunk_len": "raw_chunk_len"}),
    left_on=["source", "doc_id", "chunk_index"], right_on=["source", "doc_id", "chunk_index"],
    how="inner",
)
merged_len_check["header_overhead"] = merged_len_check["source_text_len"] - merged_len_check["raw_chunk_len"]
print(f"\nmatched {len(merged_len_check)}/{len(chroma_df)} chunks against Section 4's simulation")
print(f"mean header overhead: {merged_len_check['header_overhead'].mean():.1f} chars "
      f"(median {merged_len_check['header_overhead'].median():.0f}, "
      f"max {merged_len_check['header_overhead'].max()})")

source_text length (WITH header) stats:
count    12304.000000
mean       787.930267
std        223.996828
min         46.000000
25%        641.000000
50%        830.000000
75%        956.000000
max       1142.000000
Name: source_text_len, dtype: float64

matched 12304/12304 chunks against Section 4's simulation
mean header overhead: 57.8 chars (median 55, max 113)


In [17]:
out_csv = "chroma_embedded_dataset.csv"
chroma_df.to_csv(out_csv, index=False, encoding="utf-8-sig")

out_npy = "chroma_embeddings.npy"
np.save(out_npy, embeddings)

import os
print(f"wrote {len(chroma_df)} rows, {len(chroma_df.columns)} columns -> {out_csv} "
      f"({os.path.getsize(out_csv) / 1024 / 1024:.2f} MB)")
print(f"wrote embeddings {embeddings.shape} float32 -> {out_npy} "
      f"({os.path.getsize(out_npy) / 1024 / 1024:.2f} MB)")
print(f"\nrow order verified: chroma_embeddings.npy[i] <-> chroma_embedded_dataset.csv row i")

wrote 12304 rows, 11 columns -> chroma_embedded_dataset.csv (26.05 MB)
wrote embeddings (12304, 1024) float32 -> chroma_embeddings.npy (48.06 MB)

row order verified: chroma_embeddings.npy[i] <-> chroma_embedded_dataset.csv row i


In [18]:
# Demonstrate the CSV <-> npy split: `source_text` (the string that was
# embedded) lives in the CSV; the actual numeric embedding it produced lives
# in the .npy, at the same row index.
demo_i = 0
demo_row = chroma_df.iloc[demo_i]
demo_vector = embeddings[demo_i]

print(f"row {demo_i}: id={demo_row['id']!r}")
print(f"source_text (string, {demo_row['source_text_len']} chars): {demo_row['source_text'][:120]!r}...")
print(f"\nembedding (numeric, from chroma_embeddings.npy[{demo_i}]): shape={demo_vector.shape}, dtype={demo_vector.dtype}")
print(f"first 10 dims: {demo_vector[:10]}")
print(f"L2 norm computed from the .npy vector: {float(np.linalg.norm(demo_vector)):.6f}")
print(f"L2 norm stored in the CSV's embedding_l2_norm column: {demo_row['embedding_l2_norm']:.6f}")
assert abs(float(np.linalg.norm(demo_vector)) - demo_row["embedding_l2_norm"]) < 1e-4, "CSV/npy row alignment broken"
print("\n-> CSV row i and chroma_embeddings.npy[i] are confirmed aligned.")

row 0: id='page_5246_c0'
source_text (string, 348 chars): 'বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nশিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ই'...

embedding (numeric, from chroma_embeddings.npy[0]): shape=(1024,), dtype=float32
first 10 dims: [ 0.01534028 -0.00499048  0.01274497 -0.01428783 -0.00429519 -0.01364531
  0.01531941  0.02003078  0.01797555 -0.01422676]
L2 norm computed from the .npy vector: 1.000000
L2 norm stored in the CSV's embedding_l2_norm column: 1.000000

-> CSV row i and chroma_embeddings.npy[i] are confirmed aligned.


## 12. Lookup a chunk's embedding by id / row_key

Utility for spot-checking: given a chunk `id` (e.g. `"page_5246_c0"`) or a
`row_key` (e.g. `"page_5246"`, which returns every chunk of that page/article),
find its metadata + `source_text` from `chroma_df` and its vector from
`embeddings` (same row index in both, as established in Section 11).

Also includes a nearest-neighbor helper: since bge-m3 vectors are L2-normalized
(confirmed in Section 11 — norm is exactly 1.0), cosine similarity between two
chunks is just their dot product, so ranking neighbors is a single matrix
multiply against all 12,304 vectors. Useful for sanity-checking that the
embedding space actually groups semantically related passages.

In [19]:
def lookup_chunk(id_or_row_key: str) -> pd.DataFrame:
    """Return chroma_df rows matching an exact `id` or a `row_key` (all its chunks)."""
    matches = chroma_df[(chroma_df["id"] == id_or_row_key) | (chroma_df["row_key"] == id_or_row_key)]
    if matches.empty:
        raise KeyError(f"no chunk found for {id_or_row_key!r}")
    return matches


def show_chunk(id_or_row_key: str, preview_chars: int = 200) -> None:
    for _, row in lookup_chunk(id_or_row_key).iterrows():
        vec = embeddings[row.name]  # chroma_df's original index == embeddings row index
        print(f"id={row['id']}  book={row['book']!r}  chapter={row['chapter']!r}  "
              f"chunk_index={row['chunk_index']}  len={row['source_text_len']}")
        print(f"  text: {row['source_text'][:preview_chars]!r}...")
        print(f"  vector: shape={vec.shape} dtype={vec.dtype} first5={vec[:5]}")
        print()


# Example: a single chunk by its full id.
show_chunk("page_5246_c0")

# Example: every chunk belonging to one page (row_key has no _cN suffix).
print("all chunks of page_5246:")
show_chunk("page_5246")

id=page_5246_c0  book='শিক্ষা সাহিত্য সংস্কৃতি'  chapter='শিক্ষা'  chunk_index=0  len=348
  text: 'বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nশিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ইসলামের ধারণা বা দৃষ্টিকোণ কি, তার সংজ্ঞা ও লক্ষ্য-উদ্দেশ্য কি সে বিষয়ে কোন চূড়'...
  vector: shape=(1024,) dtype=float32 first5=[ 0.01534028 -0.00499048  0.01274497 -0.01428783 -0.00429519]

all chunks of page_5246:
id=page_5246_c0  book='শিক্ষা সাহিত্য সংস্কৃতি'  chapter='শিক্ষা'  chunk_index=0  len=348
  text: 'বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nশিক্ষা সম্পর্কে ইসলামী দৃষ্টিকোণ\n\nইসলামের শিক্ষা দর্শন বা শিক্ষা সম্পর্কে ইসলামের ধারণা বা দৃষ্টিকোণ কি, তার সংজ্ঞা ও লক্ষ্য-উদ্দেশ্য কি সে বিষয়ে কোন চূড়'...
  vector: shape=(1024,) dtype=float32 first5=[ 0.01534028 -0.00499048  0.01274497 -0.01428783 -0.00429519]

id=page_5246_c1  book='শিক্ষা সাহিত্য সংস্কৃতি'  chapter='শিক্ষা'  chunk_index=1  len=1075
  text: 'বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক

In [20]:
def nearest_neighbors(id_or_row_key: str, k: int = 5) -> pd.DataFrame:
    """Top-k most similar OTHER chunks by cosine similarity (= dot product, vectors are unit-norm)."""
    match = lookup_chunk(id_or_row_key).iloc[0]
    query_vec = embeddings[match.name]
    sims = embeddings @ query_vec
    order = np.argsort(-sims)
    order = order[order != match.name][:k]
    out = chroma_df.iloc[order][["id", "book", "chapter", "source_text"]].copy()
    out.insert(1, "similarity", sims[order])
    out["source_text"] = out["source_text"].str.slice(0, 150)
    return out.reset_index(drop=True)


print("nearest neighbors of page_5246_c0:")
nearest_neighbors("page_5246_c0")

nearest neighbors of page_5246_c0:


,id,similarity,book,chapter,source_text
0,page_5227_c0,0.835245,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,"বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nওপরের আলোচনা থেকে স্পষ্টত মনে হয়, সত্যিকারভাবে শিক্ষার কি উদ্দেশ্য ..."
1,page_5289_c1,0.826558,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nকরে তার নিজের এবং গোটা সমাজ ও জাতির জন্যে কল্যাণকরর করে গড়ে তোলা। এ...
2,page_5227_c1,0.814583,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nৎপর্য নেই- নেই কোন নিজস্ব বৈশিষ্ট্য। বর্তমানে শিক্ষা নিছক রাজনৈতিক ম...
3,page_5291_c0,0.811308,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nইসলামী শিক্ষার পাঠ্য তালিকা\n\nইসলামের দৃষ্টিতে শিক্ষার চূড়ান্ত লক্...
4,page_5258_c6,0.809536,শিক্ষা সাহিত্য সংস্কৃতি,শিক্ষা,বই: শিক্ষা সাহিত্য সংস্কৃতি\nঅধ্যায়: শিক্ষা\n\nচূড়ান্ত সত্যের মানদণ্ডে যা কিছু বিপরীত প্রমাণিত হবে তাকে যথানিয়মে ...


## Summary

- **Script mix**: pages are 92.2% dominant-Bengali, 7.3% mixed, with a mean
  3.6% Arabic-character share overall; articles are 100% Bengali, no
  Arabic. This confirms `BAAI/bge-m3` (multilingual) over
  `all-MiniLM-L6-v2` (couldn't tokenize Bengali) was the right call —
  there's no dominant-English content to design around.
- **Cross-lingual load on retrieval**: 13.8% of all chunks carry >5%
  Arabic characters (ayat/hadith inline in Bengali tafsir), 1.0% are
  majority-Arabic. These lean entirely on `bge-m3`/`bge-reranker-v2-m3`
  cross-lingual alignment rather than lexical overlap and deserve a
  dedicated eval slice.
- **Chunking**: real `chunk_text()` produces ~3 chunks/page (median 779
  chars), 1 chunk/article. 42 tiny (<50 char) chunks are isolated headings
  — cheap cosmetic cleanup, not a correctness issue.
- **Duplicate content — the actionable finding**: 6 book names exist twice
  in `books`, all pages of *every* copy are live in the Chroma index today,
  totaling ~16% of the whole vector index. 4 of the 6 pairs are byte-exact
  duplicates, contributing chunks that are 100% wasted (crowding
  `top_k`/`fetch_k` slots with redundant sources) — dedupe these first,
  the same way `scripts/clean_corpus.py` handled book203. The other 2 pairs
  (কর্মপদ্ধতি, তাফহীমুল কুরআন) have mismatched page counts and need a
  content diff before deciding, since one copy may be a partial excerpt
  rather than a true duplicate.
- **HTML / degenerate rows**: negligible — a handful of pages retain stray
  tags, zero empty or near-empty rows.